File: isegoria/src/main/java/com/eulersbridge/isegoria/vote/Signature.java
package com.eulersbridge.isegoria.vote;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.RectF;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

public class Signature extends View {
    private static final float STROKE_WIDTH_PX = 5f;
    private static final float HALF_STROKE_WIDTH_PX = STROKE_WIDTH_PX / 2;
    private final Paint paint;
    private final Path path;

    private float lastTouchX;
    private float lastTouchY;
    private final RectF dirtyRect;

    /*private LinearLayout mContent;
    private Bitmap mBitmap;
    private File mypath;*/

    public Signature(Context context, AttributeSet attrs) {
        super(context, attrs);

        paint = new Paint(Paint.ANTI_ALIAS_FLAG);
        paint.setColor(Color.BLACK);
        paint.setStyle(Paint.Style.STROKE);
        paint.setStrokeJoin(Paint.Join.ROUND);
        paint.setStrokeWidth(STROKE_WIDTH_PX);

        path = new Path();

        dirtyRect = new RectF();
    }

    /*public void save(View v) {
        Log.v(getClass().getSimpleName(), "Width: " + v.getWidth());
        Log.v(getClass().getSimpleName(), "Height: " + v.getHeight());

        if (mBitmap == null)
            mBitmap =  Bitmap.createBitmap(mContent.getWidth(), mContent.getHeight(), Bitmap.Config.RGB_565);

        Canvas canvas = new Canvas(mBitmap);
        try {
            FileOutputStream mFileOutStream = new FileOutputStream(mypath);

            v.draw(canvas);
            mBitmap.compress(Bitmap.CompressFormat.PNG, 90, mFileOutStream);
            mFileOutStream.flush();
            mFileOutStream.close();

            //String url = Images.Media.insertImage(getContentResolver(), mBitmap, "title", null);
            //Log.v(getClass().getSimpleName(),"url: " + url);

        } catch(Exception e) {
            Log.v(getClass().getSimpleName(), e.toString());
        }
    }*/

    public void clear() {
        path.reset();
        invalidate();
    }

    @Override
    protected void onDraw(Canvas canvas) {
        canvas.drawPath(path, paint);
    }

    @SuppressLint("ClickableViewAccessibility")
    @Override
    public boolean onTouchEvent(MotionEvent event) {
        float eventX = event.getX();
        float eventY = event.getY();

        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                path.moveTo(eventX, eventY);
                lastTouchX = eventX;
                lastTouchY = eventY;
                return true;

            case MotionEvent.ACTION_MOVE:
            case MotionEvent.ACTION_UP:

                resetDirtyRect(eventX, eventY);
                int historySize = event.getHistorySize();
                for (int i = 0; i < historySize; i++) {
                    float historicalX = event.getHistoricalX(i);
                    float historicalY = event.getHistoricalY(i);
                    expandDirtyRect(historicalX, historicalY);
                    path.lineTo(historicalX, historicalY);
                }

                path.lineTo(eventX, eventY);

                break;

            default:
                return false;
        }

        invalidate((int) (dirtyRect.left - HALF_STROKE_WIDTH_PX),
                (int) (dirtyRect.top - HALF_STROKE_WIDTH_PX),
                (int) (dirtyRect.right + HALF_STROKE_WIDTH_PX),
                (int) (dirtyRect.bottom + HALF_STROKE_WIDTH_PX));

        lastTouchX = eventX;
        lastTouchY = eventY;

        return true;
    }

    private void expandDirtyRect(float historicalX, float historicalY) {
        if (historicalX < dirtyRect.left) {
            dirtyRect.left = historicalX;
        } else if (historicalX > dirtyRect.right) {
            dirtyRect.right = historicalX;
        }

        if (historicalY < dirtyRect.top) {
            dirtyRect.top = historicalY;
        } else if (historicalY > dirtyRect.bottom) {
            dirtyRect.bottom = historicalY;
        }
    }

    private void resetDirtyRect(float eventX, float eventY) {
        dirtyRect.left = Math.min(lastTouchX, eventX);
        dirtyRect.right = Math.max(lastTouchX, eventX);
        dirtyRect.top = Math.min(lastTouchY, eventY);
        dirtyRect.bottom = Math.max(lastTouchY, eventY);
    }
}

File: isegoria/src/main/kotlin/com/eulersbridge/isegoria/vote/Signature.kt
package com.eulersbridge.isegoria.vote

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View

class Signature(context: Context, attrs: AttributeSet) : View(context, attrs) {

    companion object {
        private const val STROKE_WIDTH_PX = 5f
        private const val HALF_STROKE_WIDTH_PX = STROKE_WIDTH_PX / 2
    }

    operator fun MotionEvent.component1() = x
    operator fun MotionEvent.component2() = y

    private val paint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val path = Path()

    private var lastTouchX = 0.toFloat()
    private var lastTouchY = 0.toFloat()
    private val dirtyRect = RectF()

    /*private val mContent: LinearLayout? = null
    private var mBitmap: Bitmap? = null
    private val mypath: File? = null*/

    init {
        paint.apply {
            color = Color.BLACK
            style = Paint.Style.STROKE
            strokeJoin = Paint.Join.ROUND
            strokeWidth = STROKE_WIDTH_PX
        }
    }

    /*fun save(v: View) {
        Log.v(javaClass.simpleName, "Width: " + v.width)
        Log.v(javaClass.simpleName, "Height: " + v.height)

        if (mBitmap == null)
            mBitmap = Bitmap.createBitmap(mContent!!.width, mContent.height, Bitmap.Config.RGB_565)

        val canvas = Canvas(mBitmap!!)
        try {
            val mFileOutStream = FileOutputStream(mypath!!)

            v.draw(canvas)
            mBitmap!!.compress(Bitmap.CompressFormat.PNG, 90, mFileOutStream)
            mFileOutStream.flush()
            mFileOutStream.close()

            //String url = Images.Media.insertImage(getContentResolver(), mBitmap, "title", null);
            //Log.v(getClass().getSimpleName(),"url: " + url);

        } catch (e: Exception) {
            Log.v(javaClass.simpleName, e.toString())
        }
    }*/

    fun clear() {
        path.reset()
        invalidate()
    }

    override fun onDraw(canvas: Canvas) {
        canvas.drawPath(path, paint)
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(event: MotionEvent): Boolean {
        val (x, y) = event

        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                path.moveTo(x, y)
                lastTouchX = x
                lastTouchY = y
                return true
            }

            MotionEvent.ACTION_MOVE, MotionEvent.ACTION_UP -> {
                resetDirtyRect(x, y)
                val historySize = event.historySize
                for (i in 0 until historySize) {
                    val historicalX = event.getHistoricalX(i)
                    val historicalY = event.getHistoricalY(i)
                    expandDirtyRect(historicalX, historicalY)
                    path.lineTo(historicalX, historicalY)
                }

                path.lineTo(x, y)
            }

            else -> return false
        }

        invalidate(
            (dirtyRect.left - HALF_STROKE_WIDTH_PX).toInt(),
            (dirtyRect.top - HALF_STROKE_WIDTH_PX).toInt(),
            (dirtyRect.right + HALF_STROKE_WIDTH_PX).toInt(),
            (dirtyRect.bottom + HALF_STROKE_WIDTH_PX).toInt()
        )

        lastTouchX = x
        lastTouchY = y

        return true
    }

    private fun expandDirtyRect(historicalX: Float, historicalY: Float) {
        if (historicalX < dirtyRect.left) {
            dirtyRect.left = historicalX
        } else if (historicalX > dirtyRect.right) {
            dirtyRect.right = historicalX
        }

        if (historicalY < dirtyRect.top) {
            dirtyRect.top = historicalY
        } else if (historicalY > dirtyRect.bottom) {
            dirtyRect.bottom = historicalY
        }
    }

    private fun resetDirtyRect(eventX: Float, eventY: Float) {
        dirtyRect.apply {
            left = Math.min(lastTouchX, eventX)
            right = Math.max(lastTouchX, eventX)
            top = Math.min(lastTouchY, eventY)
            bottom = Math.max(lastTouchY, eventY)
        }
    }
}